File: python/nobei_core/contract.py
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator

GENERATION_SCHEMA_INVALID = "GENERATION_SCHEMA_INVALID"


def _contains_reference(value: object) -> bool:
    if isinstance(value, list):
        return any(_contains_reference(item) for item in value)
    if not isinstance(value, dict):
        return False
    return any(key == "$ref" or _contains_reference(item) for key, item in value.items())


def _json_path(path: object) -> str:
    result = "$"
    for part in path:
        result += f"[{part}]" if isinstance(part, int) else f".{part}"
    return result


@dataclass(frozen=True)
class CandidateContract:
    schema: dict[str, Any]
    schema_version: int
    schema_sha256: str
    _validator: Draft202012Validator

    def validate(self, value: object) -> list[dict[str, str]]:
        errors = sorted(
            self._validator.iter_errors(value),
            key=lambda error: (list(error.absolute_path), error.validator),
        )
        return [
            {"path": _json_path(error.absolute_path), "keyword": str(error.validator)}
            for error in errors
        ]


def load_candidate_contract(package_root: str | Path) -> CandidateContract:
    schema_path = Path(package_root) / "contracts" / "l1-candidate.schema.json"
    schema_bytes = schema_path.read_bytes()
    schema = json.loads(schema_bytes.decode("utf-8"))
    if _contains_reference(schema):
        raise ValueError("CANDIDATE_CONTRACT_REFERENCES_FORBIDDEN")
    Draft202012Validator.check_schema(schema)
    validator = Draft202012Validator(schema)
    schema_version = schema["properties"]["schemaVersion"]["const"]
    if not isinstance(schema_version, int):
        raise ValueError("CANDIDATE_CONTRACT_SCHEMA_VERSION_MISSING")
    return CandidateContract(
        schema=schema,
        schema_version=schema_version,
        schema_sha256=hashlib.sha256(schema_bytes).hexdigest(),
        _validator=validator,
    )

File: src/product/contract.ts
import { createHash } from 'node:crypto'
import { readFileSync } from 'node:fs'
import { join } from 'node:path'
import { Ajv2020 } from 'ajv/dist/2020.js'

export const GENERATION_SCHEMA_INVALID = 'GENERATION_SCHEMA_INVALID'

export interface CandidateContractValidationError {
  path: string
  keyword: string
}

export interface CandidateContract {
  schema: Record<string, unknown>
  schemaVersion: number
  schemaSha256: string
  validate(value: unknown): CandidateContractValidationError[]
}

const schemaResource = ['contracts', 'l1-candidate.schema.json']

function rejectsReferences(value: unknown): boolean {
  if (Array.isArray(value)) return value.some(rejectsReferences)
  if (value === null || typeof value !== 'object') return false
  return Object.entries(value).some(([key, child]) => key === '$ref' || rejectsReferences(child))
}

function schemaVersion(schema: Record<string, unknown>): number {
  const properties = schema.properties
  if (properties === null || typeof properties !== 'object') {
    throw new Error('CANDIDATE_CONTRACT_SCHEMA_VERSION_MISSING')
  }
  const versionProperty = (properties as Record<string, unknown>).schemaVersion
  if (versionProperty === null || typeof versionProperty !== 'object') {
    throw new Error('CANDIDATE_CONTRACT_SCHEMA_VERSION_MISSING')
  }
  const version = (versionProperty as Record<string, unknown>).const
  if (typeof version !== 'number') throw new Error('CANDIDATE_CONTRACT_SCHEMA_VERSION_MISSING')
  return version
}

export function loadCandidateContract(packageRoot: string): CandidateContract {
  const schemaPath = join(packageRoot, ...schemaResource)
  const schemaBytes = readFileSync(schemaPath)
  const schema = JSON.parse(schemaBytes.toString('utf8')) as Record<string, unknown>
  if (rejectsReferences(schema)) throw new Error('CANDIDATE_CONTRACT_REFERENCES_FORBIDDEN')

  const ajv = new Ajv2020({ allErrors: true, strict: true })
  const validator = ajv.compile(schema)

  return {
    schema,
    schemaVersion: schemaVersion(schema),
    schemaSha256: createHash('sha256').update(schemaBytes).digest('hex'),
    validate(value: unknown): CandidateContractValidationError[] {
      if (validator(value)) return []
      return (validator.errors ?? [])
        .map((error) => ({ path: error.instancePath, keyword: error.keyword }))
        .sort((left, right) => left.path.localeCompare(right.path) || left.keyword.localeCompare(right.keyword))
    },
  }
}
